### ISSUE TITLE
refactor(compiler): give operands an encoding/semantic spec instead of a bare width

### ISSUE THREAD

**Issue by gengjiawen** [gengjiawen/monkey-rust#320]
## Context

#317 added compile-time checks so oversized operands are rejected instead of being silently truncated by `make_instructions`. That closed the correctness hole, but three boundary/reporting mistakes still had to be corrected across two follow-up commits:

- index-vs-count off-by-one (the 257th local was reported as operand 256 exceeding a maximum of 255, even though 256 locals are legal)
- `OpHash` advertised twice the real pair limit, because its operand is `pairs * 2`
- free variables reported a maximum of 256 (the `OpGetFree` u8 slot) when `OpClosure`'s u8 capture *count* actually caps them at 255

The count/index width literals live in five places (four `ensure_*` helpers plus the hash special case), and the jump-target check carries a sixth, separate `u16` ceiling. There are ten semantic `ensure_*` call sites where a human has to pick the right helper — and two of the three bugs above were a `_count`/`_index` mis-pick. The operand's meaning is not written down anywhere the compiler can consult; it exists only in the head of whoever chose the helper.

## Related operand knowledge is spread across three places

Operand knowledge is currently split across three hand-written sites:

1. **compiler** — `ensure_*` calls choose a limit per emit site (`compiler/compiler.rs`)
2. **VM decode** — each arm interprets its operands (`compiler/vm.rs`, `gc/vm.rs`)
3. **snapshot validation** — `validate_instruction_stream` (`compiler/snapshot.rs:429`) applies contextual checks in a `match opcode`: `OpConst` bounds-checks the constant pool, `OpClosure` checks its constant kind, and jumps check targets

The opcode table can centralize the wire layout and context-free operand shape — index, count, scaled count, or tag — but it cannot eliminate every consumer-specific check. Constant kinds and instruction boundaries require snapshot context; local/free-slot validity requires closure or execution context. Those checks should remain explicit in their consumers.

Even among the context-free checks there is already a gap the table can close: snapshot validation does not reject an unknown `OpMethod` kind.

## Why the opcode table is the right home

`DEFINITIONS` is already the single source of truth for the wire format. `compute_abi_fingerprint()` (`compiler/snapshot.rs:86`) walks it and absorbs each opcode's value, name and operand widths to decide whether a `.mbc` snapshot is still loadable:

```rust
for &width in definition.operand_widths() {
    hash.absorb_u64(width as u64);
}
```

It just doesn't carry enough information for anyone else to use.

### 1. Split encoding from semantics

An earlier draft of this issue proposed a single `OperandKind::max_count()`. That does not work: it would have to mean three different things at once — the largest raw value the wire accepts, the largest count the user may write, and the semantic constraint on the value — and those diverge on almost every operand:

- `OpHash`'s largest raw operand is 65534 (it is always even); the user-facing pair limit is 32767
- 255 free variables means legal slot indices `0..=254`
- a jump target is a byte offset, not a count of anything
- `OpGetBuiltin`'s real range is `BuiltIns.len()`, not u8 capacity
- `OpMethod`'s kind operand is a discriminant; for two variants the legal values are `0..=1`

So keep them apart:

```rust
pub struct OperandSpec {
    pub encoding: OperandEncoding,   // U8 | U16 — what the wire can hold
    pub semantic: OperandSemantic,   // what the value means
}

pub enum OperandSemantic {
    /// Zero-based index; N items means legal values 0..=N-1.
    Index { limit: IndexLimit },
    /// The operand is the count itself.
    Count,
    /// The operand is `source * factor` (OpHash: factor 2).
    ScaledCount { factor: usize },
    /// A byte offset into the enclosing instruction stream.
    JumpTarget,
    /// Selects a variant. See #321 — a range check alone is not enough.
    Tag { variants: usize },
}
```

with separate context-free and contextual operations rather than one number:

- `validate_raw(raw) -> Result<(), _>` — width plus context-free constraints such as tag range or scaled-count divisibility
- `source_limit(context) -> Option<usize>` — how many the user may write, when that limit is meaningful
- `encode_source_value(n, context) -> Result<usize, _>` — validate and transform atomically
- `validate_in_context(raw, context) -> Result<(), _>` — table, closure, constant-kind, or instruction-boundary constraints

The encoding operation matters. `ScaledCount { factor: 2 }` must own both the limit check and a `checked_mul(2)`, not merely compute a ceiling next to a hand-written `len() * 2` at the emit site. Only then is the `OpHash` message bug structurally impossible rather than just fixed.

`IndexLimit` also needs to identify the non-width cases: `OpGetBuiltin` is bounded by `BuiltIns.len()`, and a free slot is bounded per-closure by that closure's capture count, not by its own u8 width. Those limits cannot be answered by a zero-argument spec method; the relevant validation path must receive context.

### 2. Do not break the public API

`OpcodeDefinition::operand_widths()` (`compiler/op_code.rs:24`) is `pub` and returns a borrowed `&[i32]`. Replacing the stored `Vec<i32>` with specs and "mapping over `width()`" cannot return a borrowed slice — changing the signature to `Vec<i32>` or an iterator is a breaking change for a published crate. This is the same trap the `add_constant` signature hit in #317.

Keep the method: have the constructor derive the widths from the specs once and cache them, so `operand_widths()` keeps returning `&[i32]`, and add a new `operands() -> &[OperandSpec]` alongside it.

### 3. Check operand arity in `make_instructions`

`compiler/op_code.rs:342` iterates with `zip`:

```rust
for (o, w) in operands.iter().zip(widths) {
```

Too few operands silently emits a truncated instruction; too many silently drops the extras. #317 hardened this function's *widths* with an unconditional `assert!` precisely because it is the public encoding entry point — arity belongs in the same guard:

```rust
assert_eq!(operands.len(), specs.len(), "{:?} expects {} operands, got {}", ...);
```

### 4. Fix the direction of the compiler/VM global-limit relation

`GLOBAL_SIZE` is duplicated in `compiler/vm.rs:17` and `gc/vm.rs:24`.

An earlier draft proposed `assert!(GLOBAL_SIZE <= GlobalIndex.max_count())`. That guards the harmless direction: `GLOBAL_SIZE` larger than the operand capacity only means the VM has slots the compiler can never address. The dangerous direction is the opposite — drop `GLOBAL_SIZE` to 1000 and the compiler still happily accepts 65536 globals, and the VM indexes out of bounds. The safe relation is:

```rust
const _: () = assert!(COMPILER_GLOBAL_LIMIT <= GLOBAL_SIZE);
```

or an equality assertion if the two are meant to move in lockstep. Add the invariant next to each VM's existing constant without moving the public `vm::GLOBAL_SIZE` paths as part of this refactor.

`STACK_SIZE` and `MAX_FRAMES` are also duplicated, but they are shared runtime resource policy rather than ISA operand schema. Centralizing them is a separate, lower-priority issue.

### 5. Test in three layers

An earlier draft proposed "for every operand, generate a program at `max - 1` / `max` / `max + 1` and assert it either compiles and runs correctly or returns a `CompileError`". That target is not reachable. Compile-time operand checks and runtime resource limits are independent: an array of 65,536 elements is rejected at compile time, while a *legal* 65,535-element array compiles fine and then fails at the 2,049th push under `STACK_SIZE = 2048`. So:

- **schema / encoder** — the widest raw value for each spec round-trips through `make_instructions` / `read_operands`; out-of-range values, illegal tags, odd `OpHash` counts and wrong operand arity are all rejected
- **compiler** — source-level boundaries produce the right `CompileError`, with the right number in the message
- **VM** — cross-backend (classic vs GC) result equality, only within what runtime resources actually allow

This is what turns "we fixed eight sites" into "this class is closed" — nobody writes a program with 300 locals by hand, which is why the original truncation bug survived as long as it did.

## ABI impact

For this refactor: none. `operand_widths()` keeps returning the same values, so `compute_abi_fingerprint()` is unchanged and `FORMAT_VERSION` does not need a bump — existing valid, compiler-produced `.mbc` files stay loadable.

That guarantee does **not** extend to future semantic changes. The fingerprint absorbs opcode values, names and widths only. Adding a new `Tag` variant changes no width, so an older VM would accept a newer snapshot and then misinterpret it. Whenever a tag set or a discriminant mapping changes, either a stable semantic ID has to go into the fingerprint or `FORMAT_VERSION` has to move. See #321.

## How other implementations handle this

Line references are against [Lua 5.4.6](https://github.com/lua/lua/tree/v5.4.6) and QuickJS [`04be246`](https://github.com/bellard/quickjs/tree/04be246001599f5995fa2f2d8c91a0f198d3f34c) (the `master` HEAD when this issue was filed).

|  | Lua 5.4.6 | QuickJS | Monkey (today) |
|---|---|---|---|
| **Instruction encoding** | fixed 32-bit word, bitfields (`SIZE_OP` 7, `SIZE_A/B/C` 8, `SIZE_Bx` 17, `SIZE_Ax`/`SIZE_sJ` 25) | byte stream, variable length: 1 opcode byte + 0–9 operand bytes (`with_get_var` and friends are 10 bytes total, `quickjs-opcode.h:190-194`) | byte stream, variable length: 1 opcode byte + 0–3 operand bytes |
| **Opcode count** | 83 | **244** final bytecode opcodes (178 general + 66 short variants under `#if SHORT_OPCODES`); the header declares 263 including 19 temporary opcodes that reuse short-opcode numbers — see `OP_COUNT, /* excluding temporary opcodes */` at `quickjs.c:1124` | 36 |
| **Local slot** | 8-bit `A` field; `MAXVARS` = **200** (`lparser.c:35`), deliberately below `MAXREGS` = 255 (`lcode.c:35`) to leave register headroom for temporaries | u16 `OP_get_loc`; short forms `get_loc0..3` (index baked into the opcode, 0 operand bytes) and `get_loc8` | u8 `OpGetLocal` → **256** |
| **Constant index** | 17-bit `Bx` (`OP_LOADK`) → 131 071; escapes to 25-bit `Ax` via `OP_LOADKX` + `OP_EXTRAARG` | u32 `OP_push_const` (`DEF(push_const, 5, ...)`); short `push_const8` | u16 `OpConst` → **65 536** |
| **Jump target** | 25-bit signed `sJ` (was 17-bit `sBx` in 5.3) | i32 `OP_goto` (`DEF(goto, 5, ...)`), rewritten to `goto8` / `goto16` after backpatching | u16 `OpJump` → **65 535 bytes** |
| **Call arg count** | 8-bit `B` field | u16 `npop`; short `call0..3` | u8 `OpCall` → **255** |
| **Upvalue / free variable** | `MAXUPVAL` = 255 (`lfunc.h:29`) | u16 `OP_get_var_ref`, capped by `JS_MAX_LOCAL_VARS` | u8 `OpGetFree` slot, really bounded by `OpClosure`'s u8 capture count → **255** |
| **Where the limit number comes from** | **chosen constants** — 200 and 255 are policy, not field capacity | **a chosen constant** — `JS_MAX_LOCAL_VARS` = 65 534 (`quickjs.c:210`), two sentinels below u16 | **fallout from the operand width**, with a human picking the matching helper at each check site |
| **Widening mechanism** | wide-variant opcode pair: `OP_LOADKX` + `OP_EXTRAARG` | none needed — the default form is already wide | none |
| **Narrowing mechanism** | none | post-pass rewrite (`quickjs.c:34739`): `idx < 4` → `get_loc0+idx`, `idx < 256` → `get_loc8`, else keep the wide form | none |
| **On overflow** | parser error: `too many %s (limit is %d) in %s` (`lparser.c:74`), `control structure too long` (`lcode.c:172`), `function or expression needs too many registers` (`lcode.c:471`) | `JS_ThrowInternalError`: `too many local variables` (`quickjs.c:24179`), `too many arguments` (`:24259`), `stack overflow (op=%d, pc=%d)` (`:35722`) | `CompileError`: `too many locals: 257 exceeds the maximum of 256` |
| **Encoder-level backstop** | `check_exp` → `lua_assert` → compiled out unless `LUAI_ASSERT` (debug only) | none; limits are enforced at parse time | unconditional `assert!` in `make_instructions` (#317) — widths only, arity unchecked, see §3 |

Three things worth taking from this:

**1. Both projects treat the limit as a decision, not a derivation.** Lua caps locals at 200 when the field holds 255; QuickJS caps at 65 534 when u16 holds 65 536. That is why each can be a single `#define` that never drifts. Ours genuinely *is* a function of the operand width and the operand's meaning — which is the argument for putting both in the table and deriving the limit, rather than having a human restate the pairing at every check site.

**2. Both check at the front end, not at the encoder.** So do we. Lua's encoder-level width check is debug-only and QuickJS has none at all, which makes #317's unconditional `assert!` stricter than either. That is defensible here because `make_instructions` is `pub` on a published crate, i.e. reachable by callers we cannot see — whereas Lua's and QuickJS's encoders are internal implementation, not public API. (For the record: `monkey-compiler`'s three crates.io reverse dependencies are our own workspace crates, `monkey-gc`, `monkey-asm` and `monkey-wasm`. The concern is unknown external consumers of a published crate, not three known ones.)

**3. If we ever do widen, QuickJS's direction is the better one.** Emit the wide form always and let an optional pass narrow it, rather than emitting narrow and escalating to a wide variant. The narrowing pass can be compiled out wholesale (`#if SHORT_OPCODES`) without affecting correctness, which is not true of a mandatory widening path. Note this is not a claim that narrowing is risk-free: a pass that narrows an index or a jump *wrongly* is still a miscompile. See non-goals.

## Non-goals

- **Widening any operand.** Nobody has hit these ceilings. If we ever do, follow QuickJS's shape — not `OpConstWide`-style escalation, and not a JVM `wide` prefix or LEB128, both of which would destroy the property that one opcode byte statically determines the operand layout. `read_operands`, `Instructions::string()` and `validate_instruction_stream` all depend on that property.
- **Typed decoding of `OpMethod`'s kind operand.** A `Tag` spec bounds the value's *range*; it does not make the VMs decode it exhaustively, and the failure mode there is an in-range value being misread. That is #321.
- **Centralizing VM runtime resource limits.** `STACK_SIZE` and `MAX_FRAMES` are runtime policy, not operand schema. If moved later, preserve the existing public `vm::GLOBAL_SIZE` paths through re-exports.
- **Source locations in compile errors.** That is #319.
